### ISSUE TITLE
SimplifySetCC with immediate integer problem

### ISSUE THREAD

**Issue by Nafeng** [llvm/llvm-project#56666]
SimplifySetCC should not combine the setGE into setGT when isLegalICmpImmediate return false；
this combine would cause generating a imm copy to reg instruction;
```
// IR
%cmp = icmp sge i32 %a, -128

// Debug log print
Combining: t11: i1 = setcc t2, Constant:i32<-128>, setge:ch
Creating constant: t15: i32 = Constant<-129>
Creating new node: t17: i1 = setcc t2, Constant:i32<-129>, setgt:ch
 ... into: t17: i1 = setcc t2, Constant:i32<-129>, setgt:ch
```
```
// in my TargetLoweringImpl
  bool isLegalICmpImmediate(int64_t val) const override {
    return val <= 127 && val >= -128;
  }
// In SimplifySetCC
if ((DCI.isBeforeLegalizeOps() ||
             isCondCodeLegal(NewCC, VT.getSimpleVT())) &&
            (!N1C->isOpaque() || (C.getBitWidth() <= 64 &&
                                  isLegalICmpImmediate(C.getSExtValue())))) {
          return DAG.getSetCC(dl, VT, N0,
                              DAG.getConstant(C, dl, N1.getValueType()),
                              NewCC);
        }
```
The root cause is !N1C->isOpaque() is always true;
What's a Opaque Constant? I can not find any Constant was set opaque;

**Comment by topperc** [comment]
`%cmp = icmp sge i32 %a, -128` is not canonical IR. InstCombine in the middle end will have already turned that into `%cmp = icmp sgt i32 %a, -129`

Preventing SimplifySetCC from changing sge won't solve much unless you also teach SimplifySetCC to convert `%cmp = icmp sgt i32 %a, -129` to `%cmp = icmp sge i32 %a, -128` to undo InstCombine. Both AArch64 and ARM workaround this by fixing the immediate when the setcc is converted to cmp during lowering. See `getARMCmp` and `getAArch64Cmp`

Opaque constants are constants that have been hoisted by the ConstantHoisting pass.

**Comment by s-barannikov** [comment]
@RKSimon 
Note that there are two such places, did you change both of them?

I tried several different variants, each leading to different number of failed tests. I'm certain I tried your variant, and it is possible that it did not lead to any regressions. However, this variant does not seem right since it still touches opaque constant before legalize ops. I.e. it may still duplicate "expensive" constant before op legalization.
Other variants I've tried (probably not exhaustive, I may recall later):
BeforeLegalize || (CondLegal && ConstantSmallAndLegal)
(BeforeLegalize || CondLegal) && ConstantSmallAndLegal

**Comment by s-barannikov** [comment]
@topperc 
Right, and this is not the only case. The comparison is conducted in i32 type, so there may be any i32 constant, possibly more inconvenient than -129.
Until legalizer learns to lower such comparisons to make the constant "legal" (as determined by isLegalICmpImmediate), fixing this check does not seem to make much sense. What we could do is to disable this transformation for opaque constants completely, but this leads to some regressions, too.
